### ISSUE TITLE
GeMoMa Extractor error StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by marta-fb** [Jstacs/Jstacs#8]
Hi, 

Thanks for your tool, it's been very useful so far!
I'm trying to annotate a new genome but I keep getting an error in Extractor. The code I'm using is:
`java -jar $jar CLI Extractor a=${ref_annotation} g=${ref_genome} Ambiguity=AMBIGUOUS outdir=${out} r=true sefc=true d=false`

And the error:
number of detected CDS lines: 264010
number of detected genes: 20033
number of detected transcripts: 25635
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -1
 at java.lang.String.charAt(String.java:658)
        at projects.gemoma.Extractor.transcript(Extractor.java:892)
        at projects.gemoma.Extractor.extract(Extractor.java:706)
        at projects.gemoma.Extractor.run(Extractor.java:144)
        at projects.gemoma.GeMoMaModule.run(GeMoMaModule.java:92)
        at de.jstacs.tools.ui.cli.CLI.run(CLI.java:426)
        at projects.gemoma.GeMoMa.main(GeMoMa.java:378)

Thanks,
Marta

**Comment by JensKeilwagen** [comment]
Hi Marta,

thanks a lot for your interest in GeMoMa. Which version do you use?
The problem is that line number might change with the version. My local version does not seem to have a problem in line 872 of Extractor, as it does not call charAt in this line.

best regards, Jens

bets regards, Jens

**Comment by JensKeilwagen** [comment]
I had a closer look in the proximity of this line, probably finding the place where charAt is called. Could it be that you have transcripts without amino acid sequence?

**Comment by marta-fb** [comment]
Hi Jens,

Thanks for taking a look at this.
I'm using Extractor v 1.7.1 downloaded from this link: http://www.jstacs.de/index.php/GeMoMa#Download

I'm using a GFF with refSeq annotation. I removed all lncRNAs, tRNAs and snRNA from the annotation file but I still get the error.

Thanks,
Marta

**Comment by JensKeilwagen** [comment]
Hi Marta,

lncRNA, tRNA and snRNA should not be the problem. These features are ignored while parsing.
It seems that a mRNA/transcript has an empty amino accid sequence. I think I know how to fix this.
It would be great if you could provide a link for the reference used. I could check whether I can fix it.

Best regards, Jens

**Comment by marta-fb** [comment]
Hi Jens,

Thanks!
I tried two references, and in both the same thing happens:

- https://www.ncbi.nlm.nih.gov/assembly/GCF_000001905.1/
- https://www.ncbi.nlm.nih.gov/assembly/GCF_000313985.2/

I used the refSeq genome and annotation.

Best wishes
Marta

**Comment by JensKeilwagen** [comment]
Hi Marta,

I downloaded the annotation and genome for the first organism, reproduced and identified the problem.
As assumed before the amino acid sequence is empty for the last coding exon. This can happen if the last coding exon has less than 3 coding nucleotides.  Here is an example: rna-XM_023543213.1. (In total, there were 54 exmaples in this data set!) 

In the code, I checked whether the amino acid sequence is null, but not whether it has length zero. So this is a real bug. Thanks for reporting. I fixed it and it will be available in the next GeMoMa version.

The problem occurs only if you use sefc=true. As the annotation includes the stop codon in the CDS, you don't need this option and you'll not have this problem ;)  

best regards, Jens
